### PR TITLE
Migrate diktat smoke tests to SAVE-cli mechanism

### LINKED ISSUE #1383: Migrate diktat smoke tests to SAVE-cli mechanism
1. For diktat smoke tests you need to setup the logic for SAVE-cli. 
Use `fix and warn` plugin and run diktat with save-cli. Support only config files, wrapper will be created later.

2. You should not break Unit testing of smoke tests - propose proper logic in Unit tests for removing these comments (matching SAVE-cli)

### PR DESCRIPTION
**Pull request by Cheshiriks**
### What's done:
* migrated diktat smoke tests to SAVE-cli mechanism
Closes #1383

### CODE REVIEW

**Review by nulls — APPROVED** [review]

File: diktat-rules/src/test/resources/test/smoke/save.toml [hunk 1]
```diff
@@ -1,8 +1,13 @@
 [general]
-execCmd="java -jar ktlint-snap.jar -R diktat-1.0.0-rc.2.jar"
+execCmd="java -jar ktlint -R diktat-1.1.0.jar"
```
Comment (petertrr) [inline diktat-rules/src/test/resources/test/smoke/save.toml:2]: Shouldn't we use current snapshot of diktat here?

Comment (Cheshiriks) [reply diktat-rules/src/test/resources/test/smoke/save.toml:2]: Yes. I tested on different versions. I will change the save.toml later

Comment (nulls) [reply diktat-rules/src/test/resources/test/smoke/save.toml:2]: can we change it to `diktat-*` ? does `ktlint` support it?

**Comment by akuleshov7** [comment]
Unit test for downloading of ktlint and putting diktat to a proper directory?

File: diktat-rules/src/test/resources/test/smoke/save.toml [hunk 2]
```diff
@@ -1,8 +1,13 @@
 [general]
-execCmd="java -jar ktlint-snap.jar -R diktat-1.0.0-rc.2.jar"
+execCmd="java -jar ktlint -R diktat-snap.jar"
```
Comment (akuleshov7) [inline diktat-rules/src/test/resources/test/smoke/save.toml:2]: let's rename to: diktat.jar here, it will be universal

**Review by petertrr — CHANGES_REQUESTED** [review]
@Cheshiriks please remove binaries from git, instead please either use an HTTP client in Java to download them before test or use maven plugin for downloading. Binaries should be placed in `target` directory so that they won't show up in Git and could be cleaned up by `maven clean`

**Comment by petertrr** [comment]
> > @Cheshiriks please remove binaries from git, instead please either use an HTTP client in Java to download them before test or use maven plugin for downloading. Binaries should be placed in `target` directory so that they won't show up in Git and could be cleaned up by `maven clean`
> 
> Haven't we decided that for now we will just put the Save in the project, and later we will do the download of the file?

Having a 50MB file in git slows down every fetch operation :( We can add maven configuration later, but downloading from a JVM test seems to be easy enough to implement it right away

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 3]
```diff
@@ -24,6 +26,7 @@ open class FixTestBase(protected val resourceFilePath: String,
     private val testComparatorUnit = TestComparatorUnit(resourceFilePath) { text, fileName ->
         format(ruleSetProviderRef, text, fileName, rulesConfigList, cb = cb)
     }
+    private val diktatVersion = "1.2.1-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt]: Yes, agree. Fixed

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 4]
```diff
@@ -58,6 +68,83 @@ open class FixTestBase(protected val resourceFilePath: String,
         )
     }
 
+    private fun getSaveForCurrentOs() = when {
+        System.getProperty("os.name").startsWith("Linux", ignoreCase = true) -> "save-0.3.1-linuxX64.kexe"
```
[the remark this answers is not in the document]
Comment (nulls) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:72]: and reuse it also in `downloadFile`?

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 5]
```diff
@@ -58,6 +68,83 @@ open class FixTestBase(protected val resourceFilePath: String,
...
+        System.getProperty("os.name").startsWith("Mac", ignoreCase = true) -> "save-0.3.1-macosX64.kexe"
+        System.getProperty("os.name").startsWith("Windows", ignoreCase = true) -> "save-0.3.1-mingwX64.exe"
+        else -> ""
+    }
+
+    private fun downloadFile(url: String, file: File) {
+        val httpClient = HttpClients.createDefault()
+        val request = HttpGet(url)
+        val response: CloseableHttpResponse = httpClient.execute(request)
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:81]: Yes. Thank you. Closed

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt]: Yes. Added

File: diktat-rules/pom.xml [hunk 6]
```diff
@@ -15,6 +15,7 @@
     <properties>
         <maven.compiler.source>1.8</maven.compiler.source>
         <maven.compiler.target>1.8</maven.compiler.target>
+        <apache.api.version>4.5.13</apache.api.version>
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply diktat-rules/pom.xml]: Renamed

**Review by nulls — APPROVED** [review]
approved with a small note

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 7]
```diff
@@ -58,6 +68,85 @@ open class FixTestBase(protected val resourceFilePath: String,
...
+        val fileSave = response.entity
+        fileSave?.let {
+            FileOutputStream(file).use { outstream -> fileSave.writeTo(outstream) }
+        }
+        response.close()
+        httpClient.close()
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:87]: It's a classic situation when you'll get opened resources if you get an exception before. You should wrap usages of resources in `use`

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+import com.pinterest.ktlint.core.LintError
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.Test
+
+import java.time.LocalDate
+
+import kotlin.io.path.createTempFile
+import kotlinx.serialization.builtins.ListSerializer
+
+class DiktatSaveSmokeTest : FixTestBase("test/smoke/src/main/kotlin",
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt:21]: Maybe, create a common base class with tests declarations and two implementations with different execution methods? Now we have duplication of all test methods, and it'll be hard to maintain in the future

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 9]
```diff
@@ -113,4 +202,9 @@ open class FixTestBase(protected val resourceFilePath: String,
 
         return testComparatorUnit.compareFilesFromFileSystem(expected, actual)
     }
+
+    companion object {
+        private const val KTLINT_VERSION = "0.46.1"
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:207]: Ideally, these versions should be set in `pom.xml` and passed into test as properties. Maybe not in this PR, but we should not forget to implement it in the future

Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:209]: Ok. I will create a issue for this when this pr is merge

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt (hunk 8 above)
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt]: Added base class

Comment (petertrr) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt]: It still has quite a lot of duplication. IIUC, the only difference is that in one set of tests you are using `runDiktatSmokeTest` method, and in the other `runDiktatSmokeTestWithSave`. So I would expect something like base class with test methods and abstract runner method which is overriden (seems easier, but I'm not sure if Junit will respect `@Test` annotation from base class) or - more complex - a [parameterized tests](https://www.baeldung.com/parameterized-tests-junit-5#6-method) with runner method as a parameter

Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt]: Ok. Changed

**Comment by Cheshiriks** [comment]
> Please add TearDown and TearUp mechanism. No need to download everything and copy everything for thousand of times. Check how @petertrr does it in save-cli

Yes. Added

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSmokeTestBase.kt [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        fixAndCompareBase(configFilePath, "LocalVariableWithOffsetExpected.kt", "LocalVariableWithOffsetTest.kt")
+    }
+
+    @Test
+    @Tag("DiktatRuleSetProvider")
+    fun `fix can cause long line`() {
+        fixAndCompareBase(configFilePath, "ManyLineTransformInLongLineExpected.kt", "ManyLineTransformInLongLineTest.kt")
+    }
+
+    abstract fun fixAndCompareBase(
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSmokeTestBase.kt:222]: I can be called simply `fixAndCompare`

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 11]
```diff
@@ -58,6 +59,64 @@ open class FixTestBase(protected val resourceFilePath: String,
...
+        return result
+    }
+
+    /**
+     * @param expectedPath path to file with expected result, relative to [resourceFilePath]
+     * @param testPath path to file with code that will be transformed by formatter, relative to [resourceFilePath]
+     * @param configFilePath path of diktat-analysis file
+     */
+    @Suppress("TOO_LONG_FUNCTION")
+    protected fun saveSmokeTest(
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:88]: Why is it located in `FixTestBase`? I would expect to find this method and the other one for regular smoke test in corresponding smoke test class implementations.

**Review by petertrr — APPROVED** [review]
Please move function definitions to more appropriate places, but overall lgtm

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt (hunk 11 above)
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt]: Fixed

**Review by petertrr — APPROVED** [review]

**Review by akuleshov7 — APPROVED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/indentation/Checkers.kt [hunk 12]
```diff
@@ -220,7 +220,7 @@ internal class DotCallChecker(config: IndentationConfig) : CustomIndentationChec
                 }
 
                 // we need to get indent before the first expression in calls chain
-                /*-
+                /* -
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/indentation/Checkers.kt:223]: Why?

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        expected: String,
+        test: String,
+    ) {
+        // FixMe: it is necessary to fix after the correct assembly of the Save-cli for MacOs
+        val systemName = System.getProperty("os.name")
+        if (systemName.startsWith("Mac", ignoreCase = true)) {
+            Assertions.assertTrue(true)
+        } else {
+            saveSmokeTest(config, test)
+        }
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/smoke/DiktatSaveSmokeTest.kt:31]: `@DisabledOnOs`?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/indentation/Checkers.kt (hunk 12 above)
Comment (Cheshiriks) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/indentation/Checkers.kt:223]: this is fixed diktat

Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/indentation/Checkers.kt:223]: Diktat 1.2.1? Snapshot diktat already knows this style of comments `/*-` and doesn't change them

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt [hunk 14]
```diff
@@ -130,6 +130,6 @@ open class FixTestBase(
     }
 
     companion object {
-        private const val SAVE_VERSION = "0.3.1"
+        private const val SAVE_VERSION = "0.3.2"
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt:133]: No. Companion object is not extended and cannot use class variables

[the remark this answers is not in the document]
Comment (Cheshiriks) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/util/FixTestBase.kt]: Fixed

**Review by nulls — APPROVED** [review]

**Pull request merged**